ocs_ci/deployment/terraform.py

```diff
@@ -17,31 +17,57 @@ class Terraform(object):
     Wrapper for terraform
     """
 
-    def __init__(self, path, bin_path=None):
+    def __init__(self, path, bin_path=None, state_file_path=None):
         """
         Initialize required variables needed for terraform
 
         Args:
             path (str): Path to the vSphere modules
             bin_path (str): Path to terraform binary installer
+            state_file_path (str): Path to terraform tfstate file
 
         """
         self.path = path
         self.terraform_installer = bin_path or os.path.expanduser(
             config.ENV_DATA["terraform_installer"]
         )
         self.is_directory_path_supported = False
-        self.terraform_version = config.DEPLOYMENT["terraform_version"]
-        config.ENV_DATA["terraform_state_file"] = os.path.join(
-            self.path, "terraform.tfstate"
-        )
+        self.terraform_version = Terraform.get_terraform_version()
+        if state_file_path:
+            self.state_file_path = state_file_path
+        else:
+            self.state_file_path = os.path.join(
+                config.ENV_DATA["cluster_path"], "terraform_data", "terraform.tfstate"
+            )
+        self.state_file_param = ""
+        config.ENV_DATA["terraform_state_file"] = self.state_file_path
         if version.get_semantic_version(
             self.terraform_version
         ) <= version.get_semantic_version("0.15"):
             self.is_directory_path_supported = True
-            config.ENV_DATA["terraform_state_file"] = os.path.join(
-                config.ENV_DATA["cluster_path"], "terraform_data", "terraform.tfstate"
-            )
+        else:
+            self.state_file_param = f"-state={self.state_file_path}"
+
+    @staticmethod
+    def get_terraform_version():
+        terraform_log_path = os.path.join(
+            config.ENV_DATA.get("cluster_path"), config.ENV_DATA.get("TF_LOG_FILE")
+        )
+        terraform_version = config.DEPLOYMENT["terraform_version"]
+        try:
+            with open(terraform_log_path, "r") as fd:
+                logger.debug(f"Reading terraform version from {terraform_log_path}")
+                for each_line in fd:
+                    if "Terraform version:" in each_line:
+                        terraform_version = each_line.split()[-1]
+                        logger.debug(
+                            f"Terraform version which will be use: {terraform_version}"
+                        )
+                        return terraform_version
```

Comment:
Just cosmetic note - this `return terraform_version` is probably not necessary here, as I think it will be used the return from `finally`.